sycl/plugins/unified_runtime/pi_unified_runtime.cpp

```diff
@@ -17,12 +17,62 @@ static void DieUnsupported() {
   die("Unified Runtime: functionality is not supported");
 }
 
+// Adapters may be released by piTearDown being called, or the global dtors
+// being called first. Handle releasing the adapters exactly once.
+static void releaseAdapters(std::vector<ur_adapter_handle_t> &Vec) {
+  static std::once_flag ReleaseFlag{};
+  std::call_once(ReleaseFlag, [&]() {
+    for (auto Adapter : Vec) {
+      urAdapterRelease(Adapter);
+    }
+    urTearDown(nullptr);
+  });
+}
+
+struct AdapterHolder {
+  ~AdapterHolder() { releaseAdapters(Vec); }
+  std::vector<ur_adapter_handle_t> Vec{};
+} Adapters;
+
 // All PI API interfaces are C interfaces
 extern "C" {
 __SYCL_EXPORT pi_result piPlatformsGet(pi_uint32 NumEntries,
                                        pi_platform *Platforms,
                                        pi_uint32 *NumPlatforms) {
-  return pi2ur::piPlatformsGet(NumEntries, Platforms, NumPlatforms);
+  pi_uint32 TotalNumPlatforms = 0;
+  std::vector<uint32_t> NumPlatformsPerAdapter{};
+  std::vector<ur_platform_handle_t> AllPlatforms{};
+  for (auto Adapter : Adapters.Vec) {
+    uint32_t NumPlatformsUR = 0;
+    HANDLE_ERRORS(urPlatformGet(&Adapter, 1, 0, nullptr, &NumPlatformsUR));
+    TotalNumPlatforms += NumPlatformsUR;
+    NumPlatformsPerAdapter.push_back(NumPlatformsUR);
+  }
+
+  if (NumPlatforms) {
+    *NumPlatforms = TotalNumPlatforms;
+  }
+
+  if (Platforms) {
+    AllPlatforms.resize(TotalNumPlatforms);

```

Comment:
do we need the two loops? what if we just use the second loop in line 67, and in that loop we call urPlatformGet, move it to Platforms[], breaking on std::min(NumEntries, TotalNumPlatforms)

Reply:
After looking at this again I realised that none of this logic was actually needed - the UR loader already handles iterating over all the known adapters so there's no need to query the platforms one adapter at a time. This has been replaced by a single call to `urPlatformGet`.

While investigating this I noticed a bug though - the L0 adapter was incorrect returning `UR_RESULT_ERROR_INVALID_VALUE` when no L0 platforms were found, which was causing the loader implementation of `piPlatformGet` to exit early. This meant that for non-L0 setups `SYCL_PREFER_UR` was silently failing and falling back to the normal PI path. I added a separate commit with this fix - [98e6cf6](https://github.com/intel/llvm/pull/10349/commits/98e6cf67d358f04515c55a4638ac5173ff6bbb7a) - can you check that it looks correct?